### PR TITLE
Fix inconsistency of topological sort

### LINKED ISSUE #31082: Dependency order between packages are not consistent.
**Description:**
We found out that dependency resolution between packages is not consistent between compilations while fixing https://github.com/ballerina-platform/ballerina-lang/issues/31009.

The dependency order is correct but different. 

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by nipunayf**
## Purpose
There are several different outputs for a topological sort. Thus, this risks the result being inconsistent.

Fixes #31082 

## Approach
Sort the dependencies before passing them to the topological sort at each iteration.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleDependency.java [hunk 1]
```diff
@@ -27,10 +27,13 @@
 public class ModuleDependency {
     private final PackageDependency packageDependency;
     private final ModuleId moduleId;
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleDependency.java:29]: Let's remove the moduleId if it is not needed now

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ResolvedPackageDependency.java [hunk 2]
```diff
@@ -61,4 +61,9 @@ public boolean equals(Object o) {
     public int hashCode() {
         return super.hashCode();
     }
+
+    @Override
+    public int compareTo(ResolvedPackageDependency resolvedPackageDependency) {
+        return this.packageId().toString().compareTo(resolvedPackageDependency.packageId().toString());
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ResolvedPackageDependency.java:67]: We should use the org and the name from the `resolvedPackage.descriptor()` for the comparison AFAIU

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ResolutionEngine.java [hunk 3]
```diff
@@ -565,5 +565,10 @@ public String toString() {
                     ",repo=" + pkgDesc.repository().orElse(null) + ",error=" + isError + "]";
             return pkgDesc.toString() + attr;
         }
+
+        @Override
+        public int compareTo(ResolutionEngine.DependencyNode dependencyNode) {
+            return this.pkgDesc.name().toString().compareTo(pkgDesc.name().toString());
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ResolutionEngine.java:571]: The comparison should include the `org` as well.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleCompilation.java [hunk 4]
```diff
@@ -63,41 +64,50 @@ public class ModuleCompilation {
...
-        List<ModuleId> sortedModuleIds = dependencyGraph.toTopologicallySortedList();
-        for (ModuleId sortedModuleId : sortedModuleIds) {
-            Package pkg = packageCache.getPackageOrThrow(sortedModuleId.packageId());
-            ModuleContext moduleContext = pkg.module(sortedModuleId).moduleContext();
+        List<ModuleDescriptor> sortedModuleDescriptors = dependencyGraph.toTopologicallySortedList();
+        for (ModuleDescriptor sortedModuleDescriptor : sortedModuleDescriptors) {
+            Optional<Package> pkg = packageCache.getPackage(sortedModuleDescriptor.org(),
+                    sortedModuleDescriptor.packageName(), sortedModuleDescriptor.version());
+            if (pkg.isEmpty()) {
+                throw new IllegalStateException("Cannot find a Package for the given details");
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleCompilation.java:108]: Refer https://github.com/ballerina-platform/ballerina-lang/pull/36361/files?diff=split&w=0#r885039562

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ResolutionEngine.java (hunk 3 above)
Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/ResolutionEngine.java:571]: Can we use the `pkgDesc.toString()`?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleDependency.java (hunk 1 above)
Comment (nipunayf) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleDependency.java:29]: There is one usage in the `BalaWriter` class that does not use the dependency graph. I can change it to use `descriptor().name()` as well. https://github.com/ballerina-platform/ballerina-lang/blob/e6fbef887ce946b8e1bf50bc128d97f17e5416ba/compiler/ballerina-lang/src/main/java/io/ballerina/projects/BalaWriter.java#L330

**Review by azinneera — APPROVED** [review]
LGTM

**Review by sameerajayasoma — APPROVED** [review]

**Pull request merged**